### DISCUSSION TITLE
How to correlate client information between client.connect and client.authenticate hook

### DISCUSSION THREAD

**Discussion by sushantAtAws** [emqx/emqx#7722] [Q&A] [answered]
Hi, 
I am authoring an auth plugin for emqx. We have a use case of authenticating a client based on the certificate and username. We get the cert pem when client.connect hook is called - this is where I can cache the pem information (`peercert`). Later, I get callback to client.authenticate hook which provides me the `username` (I don't want to override the username as pem). My problem is that I don't have anything unique and common in the the ConnInfo and ClientInfo to correlate the 2 calls.

My understanding is that `clientId` is not unique. ie, 2 clients can try to connect with the same `clientId`.

Is there anything available that I can use?

**Comment by zmstone** [comment, marked as answer]
The hooks are evaluated by the same erlang process, so the `self()` call should give you a unique identifier for the cache.

You may also simply call `put(my_auth_cache, Data)` in one hook callback and `get(my_auth_cache)` to retrieve it in later hook callbacks.
In this way, you don’t have to worry about cache eviction etc after the client is gone, because the data put in process dictionary simply dies together with the process.
